### PR TITLE
8306922: IR verification fails because IR dump is chopped up

### PR DESCRIPTION
**Pull request by eme64**
Some IR tests were failing because the logs were messed up (badly interleaved), due to multiple threads writing at the same time.

The issue was in `Compile::print_ideal_ir`. While we did lock the `tty`, this lock was broken because the node dumping sometimes calls into the VM, which can trigger Safepoints, which in turn can explicitly break the locks, with `ttyLocker::break_tty_lock_for_safepoint`.

What good is a lock if it can be broken? Not much. Though it is only broken if there is a Safepoint, so if we can avoid safepointing while holding the lock we should be safe. To verify that, we can write:

```
  NoSafepointVerifier nsv;
  ttyLocker ttyl;
```

The verifier triggered immediately, as expected.
And I now gather the whole dump into a separate `stringStream` first, so that all the Safepoints can happen before I even acquire the lock. Then, I acquire the log and just print the `stringStream` to `tty / xtty`, without triggering any Safepoint.

We still need to acquire a lock, else other threads may be printing also, and we get a bad interleaving in the locks.

I had to enable `dump_bfs` to print to an arbitrary stream, instead of just `tty`.

I un-problemlisted `compiler/c2/irTests/TestVectorConditionalMove.java`, and fixed a CompileCommand issue for it.

**Follow-up work**

[JDK-8310712](https://bugs.openjdk.org/browse/JDK-8310712) C2: check for broken tty locks due to SafePoint
We should in general go through all uses of `ttyLocker`, and add a `NoSafepointVerifier` to ensure no such lock is broken. Or maybe we just build the verifier into the ttyLocker?

[JDK-8310711](https://bugs.openjdk.org/browse/JDK-8310711) IR Framework: remove safepoint while printing handling

**Testing**

I ran `TestVectorConditionalMove.java` 1000x on master, just with the CompileCommand issue fixed. It triggered an IR issue 5 times (hit rate `0.005`). With the fix, it never triggers (`0.995^1000 = 0.007`). I also ran up to tier6 and stress-testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306922](https://bugs.openjdk.org/browse/JDK-8306922): IR verification fails because IR dump is chopped up (**Bug** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [bca07373](https://git.openjdk.org/jdk/pull/14591/files/bca07373576431c515f26f83fb0e8611a812a7bb)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [bca07373](https://git.openjdk.org/jdk/pull/14591/files/bca07373576431c515f26f83fb0e8611a812a7bb)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14591/head:pull/14591` \
`$ git checkout pull/14591`

Update a local copy of the PR: \
`$ git checkout pull/14591` \
`$ git pull https://git.openjdk.org/jdk.git pull/14591/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14591`

View PR using the GUI difftool: \
`$ git pr show -t 14591`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14591.diff">https://git.openjdk.org/jdk/pull/14591.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14591#issuecomment-1603861295)

### CODE REVIEW

File: src/hotspot/share/opto/compile.cpp [hunk 1]
```diff
@@ -551,28 +551,36 @@ void Compile::print_compile_messages() {
...
   // To enable tools to match it up with the compilation activity,
   // be sure to tag this tty output with the compile ID.
-  if (xtty != nullptr) {
-    xtty->head("ideal compile_id='%d'%s compile_phase='%s'",
-               compile_id(),
-               is_osr_compilation() ? " compile_kind='osr'" : "",
-               phase_name);
-  }
+
+  // Node dumping can cause a safepoint, which can break the ttyLocker.
```
Comment (chhagedorn) [inline src/hotspot/share/opto/compile.cpp:559]:
```suggestion
  // Node dumping can cause a safepoint, which can break the tty lock.
```

File: src/hotspot/share/opto/output.cpp [hunk 2]
```diff
@@ -2068,23 +2068,29 @@ void PhaseOutput::ScheduleAndBundle() {
...
+    tty->print("%s", ss.as_string());
   }
 #endif
 }
 
 #ifndef PRODUCT
 // Separated out so that it can be called directly from debugger
 void PhaseOutput::print_scheduling() {
+  print_scheduling(tty);
+}
```
Comment (chhagedorn) [inline src/hotspot/share/opto/output.cpp:2084]: New line:
```suggestion
}

```

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorConditionalMove.java [hunk 3]
```diff
@@ -46,7 +46,7 @@ public static void main(String[] args) {
         TestFramework.runWithFlags("-XX:-TieredCompilation",
                                    "-XX:+UseCMoveUnconditionally",
                                    "-XX:+UseVectorCmov",
-                                   "-XX:CompileCommand=compileonly,*.TestVectorConditionalMove.test*");
+                                   "-XX:CompileCommand=compileonly,compiler.c2.irTests.TestVectorConditionalMove::test*");
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorConditionalMove.java:49]: `compileonly` is probably not necessary but you can leave it as such.

**Review by chhagedorn — APPROVED** [review]
Looks good, thanks for cleaning this up!

I think it's a good idea to also check the other usages of `ttyLocker` separately.

Can you also file an RFE to remove the "safepoint while printing handling" in the IR framework? This is no longer needed with this patch but would exceed the scope of this fix. You can assign that to me.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorConditionalMove.java (hunk 3 above)
Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorConditionalMove.java]: Also removed the `-XX:-TieredCompilation` :)

**Review by chhagedorn — APPROVED** [review]
Thanks for the update, looks good!

File: src/hotspot/share/opto/compile.cpp [hunk 4]
```diff
@@ -551,28 +551,36 @@ void Compile::print_compile_messages() {
...
-  if (xtty != nullptr) {
-    xtty->head("ideal compile_id='%d'%s compile_phase='%s'",
-               compile_id(),
-               is_osr_compilation() ? " compile_kind='osr'" : "",
-               phase_name);
-  }
+
+  // Node dumping can cause a safepoint, which can break the tty lock.
+  // Buffer all node dumps, so that all safepoints happen before we lock.
+  stringStream ss;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.cpp:561]: Should we add a `ResourceMark rm;`?

File: src/hotspot/share/opto/output.cpp [hunk 5]
```diff
@@ -2068,23 +2068,30 @@ void PhaseOutput::ScheduleAndBundle() {
 
 #ifndef PRODUCT
   if (C->trace_opto_output()) {
-    tty->print("\n---- After ScheduleAndBundle ----\n");
-    print_scheduling();
+    // Buffer and print all at once
+    stringStream ss;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/output.cpp:2072]: Should we add a `ResourceMark rm;`?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

> Maybe we should in general go through all uses of ttyLocker, and add a NoSafepointVerifier to ensure no such lock is broken? Or maybe we just build the verifier into the ttyLocker? Should I file an RFE?

Yes, I think it would be good to file an RFE to investigate the impact of adding a NoSafepointVerifier to the ttyLocker.

**Comment by eme64** [comment]
Thanks @chhagedorn for all the discussions and help! And thanks @TobiHartmann for the review!
/integrate

**Pull request closed without merging**